.github/workflows/kani.yml

```diff
@@ -245,3 +245,75 @@ jobs:
           if-no-files-found: error
           # Aggressively short retention: we don't really need these
           retention-days: 3
+
+  perf-benchcomp:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Checkout Kani (pull request HEAD)
+        uses: actions/checkout@v3
+        with:
+          path: ./pr
+          ref: ${{ github.event.pull_request.head.sha }}
```

Comment:
What will happen when this job is triggered due to a push?

Reply:
Thanks, I've added a condition so that this only runs on `pull_request` events.

I experimented with also making it run on pushes. The trouble is that [push events](https://docs.github.com/webhooks-and-events/webhooks/webhook-events-and-payloads#push) don't have the 'before' commit that we need (i.e. the HEAD of the branch that we're trying to merge into). There is a 'before' field in that documentation, but it refers to something else---it's the commit that was previously at the tip of the pull request branch.

That is, if you post a PR to merge commit `abc123` into `main`, and then later force-push so that the PR is now trying to merge `def456` into `main`, then the push event will have `github.event.push.after == def456` and `github.event.push.before == abc123`. There is no `main` in the event, so we are unable to check it out to compare it against `def456`.

Reply:
Can you checkout `HEAD~`?

I would actually really like to run this when a release tag is created and compare with the previous release.

Reply:
By HEAD I guess you mean the tip of the pull request? I tried this, but this only does the right thing if there is only 1 commit in the pull request.

Of course, we squash when merging, so the push event for the merge commit would do the right thing. However, while the PR is open, checking out HEAD~ would not be checking against the correct base.

> I would actually really like to run this when a release tag is created and compare with the previous release.

I suggest that we have a separate workflow for this that triggers on pushing a particular tag. We would then checkout the tag, and the previous tag. For what you're asking for, I don't think HEAD~ is correct anyway, HEAD~ of a tag is just the commit before the tag in Kani because we're using a linear history.

I think having multiple different workflows that each run benchcomp is a good idea, because they would run in parallel but on different machines. If we do this, I would refactor the first steps of the job into a separate workflow---i.e. checkout two different refs in two directories---and we'd use that workflow from several different benchcomp jobs.

Reply:
If we had a separate 'release' branch where every commit was a tagged release, then `HEAD~` would have the semantics you suggest, but with a linear history I'm not aware of a symbolic way to jump from one tag to the previous (or next) one, you have to parse the log.

Reply:
By `HEAD~`, I'm talking about running this on `push` events not PRs.

For the tag workflow, we would need to find the previous / stable release.

Reply:
![Screen Shot 2023-04-14 at 18 51 57](https://user-images.githubusercontent.com/2887605/232121133-69df4186-069f-4cc0-a6f2-d46e8b52b6f9.png)
```
# This only works because releases are on their own branch so can be traversed with '~'

$ git show -s --pretty='%(describe:)' '1.27.0~'
1.26.0

$ git show -s --pretty='%(describe:)' '1.27.0~~'
1.25.0

$ git show -s --pretty='%(describe:)' '1.27.0^2'
34c30a9

$ git show -s --pretty='%(describe:)' '1.27.0^2~'
9bbb729
```

Reply:
BTW, you could create an action that takes two references that will be compared. Then each workflow will just have a logic to figure out the references and call the action.

Reply:
Thanks a lot for these comments @celinval.

I changed the job so that it does the right thing for both PRs and pushes, assuming that push events contain a single commit.
* For pull requests, benchcomp compares the HEAD of the PR with the HEAD of the branch that the PR is to be merged into. The workflow [sets `OLD_REF` and `NEW_REF` environment variables](https://github.com/model-checking/kani/actions/runs/4705590189/jobs/8346243267?pr=2370#step:3:8) with logic dependent on whether the event was a push or pull request, and uses these variables later for checking out.
* For pushes (here is an [example run](https://github.com/karkhaz/kani/actions/runs/4705582390/jobs/8346229668) in my personal fork), initially the 'old' and 'new' refs are the same, because we don't know what the 'old' ref is from the push event. Once the action clones both repositories, it checks out `HEAD~` in the 'old' one.